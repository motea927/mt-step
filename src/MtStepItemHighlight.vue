<template>
  <div class="mt-step-item-highlight" :style="positionStyle"></div>
</template>

<script setup lang="ts">
import { ref, computed, onMounted, onBeforeUnmount } from 'vue'
import { useRect } from './composables/useRect'

const props = defineProps<{
  selector: string
  hint?: {
    text: string
  }
}>()

const { element, rect } = useRect(props.selector)

const positionStyle = computed(() => {
  const { top, right, bottom, left, width, height } = rect.value
  return {
    top: `${top}px`,
    right: `${right}px`,
    bottom: `${bottom}px`,
    left: `${left}px`,
    width: `${width}px`,
    height: `${height}px`,
  }
})

const originalPosition = ref('')
const originalZIndex = ref('')
onMounted(() => {
  if (!element.value) return

  const { position, zIndex } = window.getComputedStyle(element.value)
  const newZIndex = window
    .getComputedStyle(document.querySelector('.mt-step') as Element)
    .getPropertyValue('--z-index-highlight')

  originalPosition.value = position
  originalZIndex.value = zIndex

  if (originalPosition.value === 'static') {
    element.value.style.position = 'relative'
  }
  element.value.style.zIndex = newZIndex
})

onBeforeUnmount(() => {
  if (!element.value) return
  element.value.style.position = originalPosition.value
  element.value.style.zIndex = originalZIndex.value
})
</script>

<style lang="scss">
.mt-step-item-highlight {
  display: block;
  position: absolute;
  background-color: rgb(255, 255, 255);
  z-index: var(--z-index-highlight);
  scale: 1.06;
  border-radius: 8px;
}
</style>
