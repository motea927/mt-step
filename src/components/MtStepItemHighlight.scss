.mt-step-item-highlight {
  display: block;
  position: absolute;
  background-color: rgb(255, 255, 255);
  z-index: var(--z-index-highlight);
  scale: 1.06;
  border-radius: 8px;
}
