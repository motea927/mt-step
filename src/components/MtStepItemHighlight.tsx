import {
  defineComponent,
  computed,
  ref,
  onMounted,
  onBeforeUnmount,
  h,
} from 'vue-demi'
import type { PropType } from 'vue-demi'
import { useRect } from '@/composables/useRect'
import './MtStepItemHighlight.scss'

const MtStepItemHighlight = defineComponent({
  props: {
    selector: {
      type: String as PropType<string>,
      required: true,
    },
    isPointerEventsDisabled: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
  },
  setup(props) {
    const { element, rect } = useRect(props.selector)

    const positionStyle = computed(() => {
      const { top, right, bottom, left, width, height } = rect.value
      return {
        top: `${top}px`,
        right: `${right}px`,
        bottom: `${bottom}px`,
        left: `${left}px`,
        width: `${width}px`,
        height: `${height}px`,
      }
    })

    const originalPosition = ref('')
    const originalZIndex = ref('')
    const originalPointerEvents = ref('')

    onMounted(() => {
      if (!element.value) return

      const { position, zIndex, pointerEvents } = window.getComputedStyle(
        element.value
      )

      const newZIndex = window
        .getComputedStyle(document.querySelector('.mt-step') as Element)
        .getPropertyValue('--z-index-highlight')

      originalPosition.value = position
      originalZIndex.value = zIndex
      originalPointerEvents.value = pointerEvents

      if (!originalPosition.value || originalPosition.value === 'static') {
        element.value.style.position = 'relative'
      }
      element.value.style.zIndex = newZIndex

      if (props.isPointerEventsDisabled) {
        element.value.style.pointerEvents = 'none'
      }
    })

    onBeforeUnmount(() => {
      if (!element.value) return
      element.value.style.position = originalPosition.value
      element.value.style.zIndex = originalZIndex.value
      element.value.style.pointerEvents = originalPointerEvents.value
    })

    return () =>
      h('div', {
        class: 'mt-step-item-highlight',
        style: positionStyle.value,
      })
  },
})

export default MtStepItemHighlight
