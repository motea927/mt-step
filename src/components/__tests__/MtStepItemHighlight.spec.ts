import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import { mount } from '@vue/test-utils'
import MtStepItemHighlight from '@/components/MtStepItemHighlight'

const mockZIndex = '799'
const mountElement = (isPointerEventsDisabled?: boolean) =>
  mount(MtStepItemHighlight, {
    propsData: {
      selector: '#test',
      ...(isPointerEventsDisabled && { isPointerEventsDisabled }),
    },
  })
describe('MtStepItemHighlight', () => {
  beforeEach(() => {
    // Create a target element for testing

    const targetElement = document.createElement('div')
    targetElement.id = 'test'
    document.body.appendChild(targetElement)
    targetElement.style.position = ''
    targetElement.style.pointerEvents = ''

    const mtStepELement = document.createElement('div')
    mtStepELement.className = 'mt-step'
    mtStepELement.style.setProperty('--z-index-highlight', mockZIndex)
    document.body.appendChild(mtStepELement)
  })

  afterEach(() => {
    // Remove the target element after testing
    ;['#test', '.mt-step'].forEach((selector) => {
      const targetElement = document.querySelector(selector)
      document.body.removeChild(targetElement as Element)
    })
  })

  it('renders correctly with rect data', () => {
    const wrapper = mountElement()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('updates position and z-index of target element', async () => {
    const wrapper = mountElement()

    const targetElement = document.querySelector('#test') as HTMLElement
    expect(targetElement.style.position).toBe('relative')
    expect(targetElement.style.zIndex).toBe(mockZIndex)
    wrapper.unmount()

    expect(targetElement.style.position).toBe('')
    expect(targetElement.style.zIndex).toBe('')
  })

  it("updates target element's pointer-event when isPointerEventsDisabled", async () => {
    const wrapper = mountElement(true)

    const targetElement = document.querySelector('#test') as HTMLElement
    expect(targetElement.style.pointerEvents).toBe('none')
    wrapper.unmount()
    expect(targetElement.style.pointerEvents).toBe('')
  })

  it('default isPointerEventsDisabled is false', async () => {
    mountElement()
    const targetElement = document.querySelector('#test') as HTMLElement
    expect(targetElement.style.pointerEvents).toBe('')
  })
})
