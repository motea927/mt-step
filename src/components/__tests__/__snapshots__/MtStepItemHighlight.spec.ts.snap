// Vitest Snapshot v1

exports[`MtStepItemHighlight > renders correctly with rect data 1`] = `
<div
  class="mt-step-item-highlight"
  style="top: 0px; right: 0px; bottom: 0px; left: 0px; width: 0px; height: 0px;"
/>
`;
